### PR TITLE
Imx477 long exposure control

### PR DESCRIPTION
**Pull request by naushir**

### CODE REVIEW

File: drivers/media/i2c/imx477.c [hunk 1]
```diff
@@ -1316,7 +1352,8 @@ static int imx477_set_ctrl(struct v4l2_ctrl *ctrl)
 		break;
 	case V4L2_CID_EXPOSURE:
 		ret = imx477_write_reg(imx477, IMX477_REG_EXPOSURE,
-				       IMX477_REG_VALUE_16BIT, ctrl->val);
+				       IMX477_REG_VALUE_16BIT, ctrl->val /
+						(1 << imx477->long_exp_shift));
```
Comment (6by9) [inline drivers/media/i2c/imx477.c]: ```(ctrl->val >> imx477->long_exp_shift)```?

File: drivers/media/i2c/imx477.c [hunk 2]
```diff
@@ -1285,20 +1292,49 @@ static void imx477_adjust_exposure_range(struct imx477 *imx477,
...
+	imx477->long_exp_shift = 0;
+
+	while (val > IMX477_FRAME_LENGTH_MAX) {
+		if (++imx477->long_exp_shift > IMX477_LONG_EXP_SHIFT_MAX) {
+			imx477->long_exp_shift = IMX477_LONG_EXP_SHIFT_MAX;
+			val = IMX477_FRAME_LENGTH_MAX;
+			break;
+		}
+
+		val /= 2;
```
Comment (6by9) [inline drivers/media/i2c/imx477.c]: ```val >>= 1```?
You're dealing with shifts so it feels odd mixing shifts and divides (even if they are the same thing).

File: drivers/media/i2c/imx477.c [hunk 3]
```diff
@@ -1285,20 +1292,49 @@ static void imx477_adjust_exposure_range(struct imx477 *imx477,
...
 }
 
+static int imx477_set_frame_length(struct imx477 *imx477, unsigned int val)
+{
+	int ret = 0;
+
+	imx477->long_exp_shift = 0;
+
+	while (val > IMX477_FRAME_LENGTH_MAX) {
+		if (++imx477->long_exp_shift > IMX477_LONG_EXP_SHIFT_MAX) {
```
Comment (6by9) [inline drivers/media/i2c/imx477.c]: Pre/post increments/decrements in logical statements always make me have to think harder than I'd like over the actual logic being checked. Separate them into two lines?

File: drivers/media/i2c/imx477.c [hunk 4]
```diff
@@ -1285,20 +1292,49 @@ static void imx477_adjust_exposure_range(struct imx477 *imx477,
...
+			imx477->long_exp_shift = IMX477_LONG_EXP_SHIFT_MAX;
```
Comment (6by9) [inline drivers/media/i2c/imx477.c]: And a second question, doesn't this mean that the range on VBLANK has been set wrongly if we exceed the limit? Or are there rounding errors?

File: drivers/media/i2c/imx477.c [hunk 5]
```diff
@@ -1927,7 +1967,8 @@ static int imx477_init_controls(struct imx477 *imx477)
 	imx477->exposure = v4l2_ctrl_new_std(ctrl_hdlr, &imx477_ctrl_ops,
 					     V4L2_CID_EXPOSURE,
 					     IMX477_EXPOSURE_MIN,
-					     IMX477_EXPOSURE_MAX,
+					     IMX477_EXPOSURE_MAX *
+					       (1 << IMX477_LONG_EXP_SHIFT_MAX),
```
Comment (6by9) [inline drivers/media/i2c/imx477.c]: I haven't compiled or run this, but doesn't this leave a max exposure that can't be achieved under most conditions. Shouldn't the exposure range (and step?) be updated when a VBLANK change results in a change of the multiplier?

Ah we do adjust it via ```imx477_adjust_exposure_range```, and that is called at the end of this function. Seeing as the default VBLANK is going to have a multiplier of x1 I'd be tempted to leave this alone, even if it is then updated at the end of the function.

File: drivers/media/i2c/imx477.c (hunk 1 above)
Comment (naushir) [reply drivers/media/i2c/imx477.c]: I purposely did it with a divide, but in hindsight, the intention is not very clear :)
Will fix this an use shifts for all the other calculations.

File: drivers/media/i2c/imx477.c (hunk 4 above)
Comment (naushir) [reply drivers/media/i2c/imx477.c]: Yes, strictly speaking, we cannot go over SHIFT_MAX based on the VBLANK limits set.  I was just being a bit defensive with this extra check.  I'll remove it.

File: drivers/media/i2c/imx477.c (hunk 5 above)
Comment (naushir) [reply drivers/media/i2c/imx477.c]: Yes, it will be updated correctly so no harm in leaving this as the orginal.

**Comment by naushir** [comment]
@pelwell and @6by9 this is ready to go in if you are both happy.

**Pull request merged**